go/vt/schema/online_ddl.go

```diff
@@ -68,42 +76,126 @@ func validateWalk(node sqlparser.SQLNode, allowForeignKeys bool) (kontinue bool,
 	return false, nil
 }
 
-// OnlineDDLStatus is an indicator to a online DDL status
-type OnlineDDLStatus string
-
-const (
-	OnlineDDLStatusRequested OnlineDDLStatus = "requested"
-	OnlineDDLStatusCancelled OnlineDDLStatus = "cancelled"
-	OnlineDDLStatusQueued    OnlineDDLStatus = "queued"
-	OnlineDDLStatusReady     OnlineDDLStatus = "ready"
-	OnlineDDLStatusRunning   OnlineDDLStatus = "running"
-	OnlineDDLStatusComplete  OnlineDDLStatus = "complete"
-	OnlineDDLStatusFailed    OnlineDDLStatus = "failed"
-)
-
 // OnlineDDL encapsulates the relevant information in an online schema change request
 type OnlineDDL struct {
-	Keyspace string      `json:"keyspace,omitempty"`
-	Table    string      `json:"table,omitempty"`
-	Schema   string      `json:"schema,omitempty"`
-	SQL      string      `json:"sql,omitempty"`
-	UUID     string      `json:"uuid,omitempty"`
-	Strategy DDLStrategy `json:"strategy,omitempty"`
-	Options  string      `json:"options,omitempty"`
+	// m is used to provide concurrent access to ReadyToComplete logic. Callers
+	// should use the IsReadyToComplete and MarkReadyToComplete methods to
+	// safely modify the value of that field rather than accessing it through
+	// the embedded protobuf message directly.
+	m sync.Mutex
+	*tabletmanagerdatapb.OnlineDDL
+
 	// Stateful fields:
-	MigrationContext   string          `json:"context,omitempty"`
-	Status             OnlineDDLStatus `json:"status,omitempty"`
-	TabletAlias        string          `json:"tablet,omitempty"`
-	Retries            int64           `json:"retries,omitempty"`
-	ReadyToComplete    int64           `json:"ready_to_complete,omitempty"`
-	WasReadyToComplete int64           `json:"was_ready_to_complete,omitempty"`
+	MigrationContext string `json:"context,omitempty" mapstructure:"context,omitempty"`
+	TabletAlias      string `json:"tablet,omitempty" mapstructure:"tablet,omitempty"`
 }
 
 // FromJSON creates an OnlineDDL from json
-func FromJSON(bytes []byte) (*OnlineDDL, error) {
-	onlineDDL := &OnlineDDL{}
-	err := json.Unmarshal(bytes, onlineDDL)
-	return onlineDDL, err
+func FromJSON(data []byte) (*OnlineDDL, error) {
```

Comment:
Will it not be safer to create a temporary `struct` in the likeness of the old `OnlineDDL` struct, read the JSON into that struct, then copy over the results to the "new" OnlineDDL fields?